storage/src/main/java/org/apache/kafka/storage/internals/log/UnifiedLog.java

```diff
@@ -206,4 +223,119 @@ public static Optional<CompletedTxn> updateProducers(ProducerStateManager produc
+            }
+
+            public void recordInvalidSequence() {
+                allTopicsStats.invalidOffsetOrSequenceRecordsPerSec().mark();
+            }
+
+            public void recordInvalidChecksums() {
+                allTopicsStats.invalidMessageCrcRecordsPerSec().mark();
+            }
+
+            public void recordNoKeyCompactedTopic() {
+                allTopicsStats.noKeyCompactedTopicRecordsPerSec().mark();
+            }
+        };
+    }
+
+    /**
+     * If the recordVersion is >= RecordVersion.V2, create a new LeaderEpochFileCache instance.
+     * Loading the epoch entries from the backing checkpoint file or the provided currentCache if not empty.
+     * Otherwise, the message format is considered incompatible and the existing LeaderEpoch file
+     * is deleted.
+     *
+     * @param dir                  The directory in which the log will reside
+     * @param topicPartition       The topic partition
+     * @param logDirFailureChannel The LogDirFailureChannel to asynchronously handle log dir failure
+     * @param recordVersion        The record version
+     * @param logPrefix            The logging prefix
+     * @param currentCache         The current LeaderEpochFileCache instance (if any)
+     * @param scheduler            The scheduler for executing asynchronous tasks
+     * @return The new LeaderEpochFileCache instance (if created), empty otherwise
+     */
+    public static Optional<LeaderEpochFileCache> maybeCreateLeaderEpochCache(File dir,
+                                                                             TopicPartition topicPartition,
+                                                                             LogDirFailureChannel logDirFailureChannel,
+                                                                             RecordVersion recordVersion,
+                                                                             String logPrefix,
+                                                                             Optional<LeaderEpochFileCache> currentCache,
+                                                                             Scheduler scheduler) throws IOException {
+        File leaderEpochFile = LeaderEpochCheckpointFile.newFile(dir);
+
+        if (recordVersion.precedes(RecordVersion.V2)) {
+            if (leaderEpochFile.exists()) {
+                LOG.warn("{}Deleting non-empty leader epoch cache due to incompatible message format {}", logPrefix, recordVersion);
+            }
+            Files.deleteIfExists(leaderEpochFile.toPath());
+            return Optional.empty();
+        } else {
+            LeaderEpochCheckpointFile checkpointFile = new LeaderEpochCheckpointFile(leaderEpochFile, logDirFailureChannel);
+            return Optional.of(currentCache.map(cache -> cache.withCheckpoint(checkpointFile))
+                    .orElse(new LeaderEpochFileCache(topicPartition, checkpointFile, scheduler)));
```

Comment:
Could you please use `orElseGet` to avoid unnecessary object creation?

Reply:
Good point, done